docs/books/valhalla-mjolnir/src/03_graph_construction_process.md

```diff
@@ -0,0 +1,310 @@
+- Simple (from one edge to another)
+- Complex (involving multiple via edges or nodes)
+- Time-dependent (only active during certain hours)
+- Mode-specific (only applying to certain vehicle types)
+
+## Hierarchical Graph Building
+
+Valhalla's multi-level graph is created by the `HierarchyBuilder`:
+
+```cpp
+// From valhalla/mjolnir/hierarchybuilder.h
+class HierarchyBuilder {
+public:
+  static void Build(const boost::property_tree::ptree& pt);
+};
+```
+
+The process involves:
+
+1. **Edge Classification**: Determining which edges belong to which hierarchy level
+2. **Shortcut Creation**: Creating direct connections between important nodes to bypass less important roads
+3. **Transition Edge Creation**: Adding special edges to connect between hierarchy levels
+
+Here's a simplified version of the hierarchy building process:
+
+```cpp
+// From src/mjolnir/hierarchybuilder.cc
+void HierarchyBuilder::Build(const boost::property_tree::ptree& pt) {
+  // Create level 0 (local) graph from the enhanced base graph
+  CreateLevel0(pt);
+  
+  // Create level 1 (regional) graph
+  CreateLevel1(pt);
+  
+  // Create level 2 (highway) graph
+  CreateLevel2(pt);
+  
+  // Add transitions between levels
+  AddTransitions(pt);
+}
+
+void HierarchyBuilder::CreateLevel1(const boost::property_tree::ptree& pt) {
+  // For each tile in level 0
+  for (const auto& tile_id : level0_tile_ids) {
+    GraphTileBuilder level1_builder;
+    
+    // Get the level 0 tile
+    const GraphTile* tile = reader.GetGraphTile(tile_id);
+    
+    // For each edge in the level 0 tile
+    for (const auto& edge : tile->directededges()) {
+      // If edge is important enough for level 1
+      if (ShouldIncludeInLevel1(edge)) {
+        // Add to level 1 graph
+        level1_builder.AddEdge(edge);
+      }
+    }
+    
+    // Create shortcuts for level 1
+    CreateShortcuts(level1_builder);
+    
+    // Save the level 1 tile
+    level1_builder.StoreTileData();
+  }
+}
+```
+
+The hierarchical structure is what enables Valhalla to efficiently route over long distances. By using higher-level tiles with shortcuts, the routing algorithm can quickly traverse large areas without examining every local road.
+
+## Shortcut Building
+
+Shortcuts are special edges that represent multiple underlying edges, allowing for faster routing on higher hierarchy levels. They're created by the `ShortcutBuilder`:
+
+```cpp
+// From valhalla/mjolnir/shortcutbuilder.h
+class ShortcutBuilder {
+public:
+  static void Build(const boost::property_tree::ptree& pt);
+};
+```
+
+The implementation in `src/mjolnir/shortcutbuilder.cc` identifies paths that can be represented as shortcuts and creates the corresponding edges. Shortcuts preserve important routing properties like:
+- Total distance
+- Travel time
+- Access restrictions
+- Turn restrictions at endpoints
```

Comment:
indeed important stuff liek that is a place where we cannot do the contraction and so shortcuts begin/end where restrictions begin/end